framework/wazuh/agent.py

```diff
@@ -676,55 +680,124 @@ def get_outdated_agents(agent_list=None, offset=0, limit=common.database_limit,
-    :return: Upgrade message.
-    """
-    # We access unique agent_id from list, this may change if and when we decide to add option to upgrade a list of
-    # agents
-    agent_id = agent_list[0]
-
-    return Agent(agent_id).upgrade(wpk_repo=wpk_repo, version=version, force=True if int(force) == 1 else False,
-                                   chunk_size=chunk_size, use_http=use_http)
-
+@expose_resources(actions=["agent:upgrade"], resources=["agent:id:{agent_list}"],
+                  post_proc_kwargs={'exclude_codes': [1703, 1818]})
+def upgrade_agents(agent_list=None, wpk_repo=None, version=None, force=False, use_http=False,
+                   file_path=None, installer=None):
+    """Start the agent upgrade process.
 
-@expose_resources(actions=["agent:upgrade"], resources=["agent:id:{agent_list}"], post_proc_func=None)
-def get_upgrade_result(agent_list=None, timeout=3):
-    """Read upgrade result output from agent.
+    Parameters
+    ----------
+    agent_list : list
+        List of agents ID's.
+    wpk_repo : str
+        URL for WPK download.
+    version : str
+        Version to upgrade to.
+    force : bool
+        force the update even if it is a downgrade.
+    use_http : bool
+        False for HTTPS protocol, True for HTTP protocol.
+    file_path : str
+        Path to the installation file.
+    installer : str
+        Selected installer.
 
-    :param agent_list: List of agents ID's.
-    :param timeout: Maximum time for the call to be considered failed.
-    :return: Upgrade result.
+    Returns
+    -------
+    ID of created tasks
     """
-    # We access unique agent_id from list, this may change if and when we decide to add option to upgrade a list of
-    # agents
-    agent_id = agent_list[0]
+    result = AffectedItemsWazuhResult(all_msg='All upgrade tasks have been created',
+                                      some_msg='Some upgrade tasks have been created',
+                                      none_msg='No upgrade task has been created',
+                                      sort_fields=['task_id'], sort_ascending='True')
+
+    agent_list = list(map(int, agents_padding(result=result, agent_list=agent_list)))
+    if not file_path:
+        wpk_repo = wpk_repo if wpk_repo else common.wpk_repo_url_4_x
+    if version and not version.startswith('v'):
+        version = f'v{version}'
+    msg = {'version': 1,
+           'origin': {'module': 'api'},
+           'command': 'upgrade' if not (installer or file_path) else 'upgrade_custom',
+           'parameters': {
+               'agents': list(),
+               'version': version,
+               'force_upgrade': force,
+               'use_http': use_http,
+               'wpk_repo': f'{wpk_repo}/' if wpk_repo and not wpk_repo.endswith('/') else wpk_repo,
+               'file_path': file_path,
+               'installer': installer
+           }
+           }
+
+    msg['parameters'] = {k: v for k, v in msg['parameters'].items() if v is not None}
```

Comment:
Same as before, messy, not generic way to remove nones.

Reply:
I left that line in order to avoid a circular import.